### PR TITLE
Add | operator

### PR DESCRIPTION
**Pull request by johnynek**
Adds a synonym for orElse, which is common in parsing libraries.

### CODE REVIEW

**Comment by johnynek** [comment]
@martijnhoekstra @regadas 

Hey folks, take a look at this. My anxiety about having two orElse methods seemed to have some truth to it... It looks like on 2.12, you can't mix a Parser0 and a Parser with orElse (or `|`). It seems that it works if both arguments are the same type.

So, to me, that suggests we should add orElse0 for Parser0 (removing orElse on Parser0), and then make | the operator only work when both arguments are Parser.

What do you think?

**Comment by martijnhoekstra** [comment]
That's really sad. Is it a bug in the overload resolution of 2.12 for abstract types, or does it happen with all types? I'll poke at it a bit this afternoon.

**Comment by regadas** [comment]
Yea, from a quick look this seems to be an issue with the dependent type in `GenT`. 2.12 seems to have issues with `p2.A >: p1.A`

```scala
    forAll(ParserGen.gen, ParserGen.gen0) { (p1, p2) =>
      val asd = p1.fa.orElse(p2.fa.asInstanceOf[Parser0[p1.A]])
      assertEquals(asd, p1.fa | p2.fa.asInstanceOf[Parser0[p1.A]])
    }
```

regular stuff works fine
```scala
Parser.string("a") | Parser.string0("b")
```

**Comment by regadas** [comment]
Given that this is 2.12 specific and "normal" use cases are fine, I'm wondering if this issue is worth the extra API juggling and `|` available only on `Parser`.

**Comment by johnynek** [comment]
good catch @regadas -- this is really about those generators not the Parser0/Parser types. If I ascribe a type, it seems to work.

So, I think this should probably be okay to merge.

**Comment by johnynek** [comment]
okay, this seems to pass and i guess was just a bug with the path dependent type.

Any objection to merge?

File: core/shared/src/test/scala/cats/parse/ParserTest.scala [hunk 1]
```diff
@@ -1193,6 +1193,29 @@ class ParserTest extends munit.ScalaCheckSuite {
...
+      assertEquals(pa.orElse(pb), pa | pb)
+    }
+    forAll(ParserGen.gen0, ParserGen.gen0) { (p1, p2) =>
+      val pa: Parser0[Any] = p1.fa
+      val pb: Parser0[Any] = p2.fa
+      assertEquals(pa.orElse(pb), pa | pb)
+    }
+    forAll(ParserGen.gen, ParserGen.gen0) { (p1, p2) =>
+      val pa: Parser[Any] = p1.fa
+      val pb: Parser0[Any] = p2.fa
```
Comment (regadas) [inline core/shared/src/test/scala/cats/parse/ParserTest.scala:1209]: yea the ascription to Any makes sense.

**Review by regadas — APPROVED** [review]
👍

**Pull request merged**